### DISCUSSION TITLE
Battery Measurement Unstable results. Example Read vcc from megatiny

### DISCUSSION THREAD

**Discussion by Brendy1986** [SpenceKonde/megaTinyCore#1020] [Q&A]
Hello Attiny1626
I used the example of measuring batteries. I would like to measure the 18650 battery 3 - 4.2V.
I connected it from the computer for the test.
The voltage on the port monitor is higher than 5.08.
I have very variable results
How to calibrate it to get precise results? What do I need to change in the code?

```
#include <megaTinyCore.h>
/* This demonstrates both how the routines to measure temperature are implemented, with explanitary serial prints
 * and then calls the library functions provided that do the same thing (just faster and with less flash usage,
 * particularly on 2-series parts, where a shortcut was found in the mathematics
 */

#define RESULTCOUNT 4
int16_t results[RESULTCOUNT];
int32_t sum;
int16_t average;

void setup() {
  // put your setup code here, to run once:
  delay(1000);
  Serial.begin(57600);
  #if MEGATINYCORE_SERIES == 2
  Serial.println("On some parts, these results may be off by more than twice the spec'ed +/- 5 degrees.");
  #endif
}

uint16_t readSupplyVoltageExplained() { // returns value in millivolts to avoid floating point
  #if MEGATINYCORE_SERIES!=2
  analogReference(VDD);
  VREF.CTRLA = VREF_ADC0REFSEL_1V5_gc;
  // there is a settling time between when reference is turned on, and when it becomes valid.
  // since the reference is normally turned on only when it is requested, this virtually guarantees
  // that the first reading will be garbage; subsequent readings taken immediately after will be fine.
  // VREF.CTRLB|=VREF_ADC0REFEN_bm;
  // delay(10);
  uint16_t reading = analogRead(ADC_INTREF);
  Serial.print(reading);
  Serial.println(" (discarded)");
  reading = analogRead(ADC_INTREF);
  Serial.println(reading);
  uint32_t intermediate = 1023UL * 1500;
  reading = intermediate / reading;
  return reading;
  #else
  analogReference(INTERNAL1V024);
  //Serial.print(analogReadEnh(ADC_VDDDIV10, 12)); // throwaway reading just for good measure.
  //Serial.println(" (discarded)");
  int32_t vddmeasure = analogReadEnh(ADC_VDDDIV10, 12); // Take it at 12 bits
 // Serial.println(vddmeasure);
  vddmeasure *= 10; // since we measured 1/10th VDD
  int16_t returnval = vddmeasure >> 2; // divide by 4 to get into millivolts.
  /* The above method of doing the math is bad, because multiplication of a 32-bit value takes
   * 81 clocks. The smart way to multiply by 10 and then divide by four is to instead take the sum
   * of the original value leftshifted one place, and the original value rightshifted one place.
   * And here most of this can be donewith a 16-bit variable not a 32 bit one
   */
  if (vddmeasure & 0x02) {
    // if last two digits were 0b11 or 0b10 we should round up
    returnval++;
  }
  return returnval;
  #endif
}
void printRegisters() {
  #if MEGATINYCORE_SERIES!=2
/*  Serial.print("ADC0.MUXPOS: ");
  Serial.printHex(ADC0.MUXPOS);
  Serial.print("  ADC0.CTRLC: ");
  Serial.printHex(ADC0.CTRLC);
  Serial.print("  VREF.CTRLA: ");
  Serial.printHex(VREF.CTRLA);
  Serial.println(); */
  #endif
}



void loop() {
  int16_t reading = readSupplyVoltageExplained();
  //Serial.print("System voltage is: ");
  //Serial.print(reading);
  //Serial.println(" mV");

  delay(1000);
  //Serial.println("Repeating above without the explanations");
  reading = readSupplyVoltage();
  float volt = reading/1000.0;
  delay(1000);

  //Serial.print("System voltage is: ");
  //Serial.print(reading);
  //Serial.println(" mV");
  Serial.print(volt);
  Serial.println(" V");

}
```

`5.38 V
5.39 V
5.37 V
5.38 V
5.36 V
5.37 V
5.36 V
5.41 V
5.38 V
5.38 V
5.40 V
5.38 V
5.42 V
5.39 V
5.41 V
5.36 V
5.37 V
5.38 V
5.41 V
5.38 V
5.37 V
5.36 V
5.37 V
5.36 V
5.38 V
5.42 V
5.46 V
5.47 V
5.44 V
5.47 V
5.42 V
`

**Comment by Modac** [comment]
Am I understanding correctly, that you are currently measuring the 5V USB voltage with the attiny? If so the variability you are seeing could just be the actual voltage. Have you checked the voltage with a multimeter?

**Reply by Brendy1986** [reply]
I made a big mistake. USB sends a lot of noise.
I connected it this way by adding a capacitor and the results are much better.
I measured the 18650 3.88V battery with a meter.
On the port monitor my results.
![sche](https://github.com/SpenceKonde/megaTinyCore/assets/118827809/1e6252f8-223e-4d2a-8996-331214224edf)

```
3.89 V
3.90 V
3.89 V
3.89 V
3.89 V
3.89 V
3.90 V
3.89 V
3.89 V
3.89 V
3.89 V
3.89 V
3.90 V
3.89 V
3.89 V
3.89 V
3.90 V
3.90 V
3.90 V
3.89 V
3.90 V
3.89 V
3.89 V
3.89 V
3.89 V
3.90 V
3.89 V
3.89 V
3.89 V
3.89 V
3.89 V
3.89 V
3.89 V
```

**Comment by Brendy1986** [comment]
I did some tests using a 3-4.2V linear stabilizer and it looks very good.
guys
Can I measure up to 12.90V without using dividers?

**Comment by hmeijdam** [comment]
No, you will damage a pin if you go over VCC + 0.5V. Above that the ESD protection diodes will begin conducting and will fry.

![image](https://github.com/SpenceKonde/megaTinyCore/assets/9595465/486b2835-5072-4e87-9df2-0311642dcb07)

**Reply by SpenceKonde** [reply]
A lot of USB ports put out a few extra tenths of a volt, especially USB chargers, to improve the charge rate of power-hungry cellphones - with 5v, you can afford to lose 0.8v at most in both the charging circuit and the cable before you're at the limit of how fast you can charge. Even the devices that use a switching converter to charge the battery will still need like a third of a volt of headroom to get you the target output voltage. So what's the problem? You need 4.5ish and you have 5! But that doesn't consider the voltage drop in the USB cable. But many micro USB cables have non-negligible resistance because copper is expensive, such that the limiting speed of charging is poor. So people would complain and give bad reviews to chargers that put out the spec'ed 5v because they charged their phones slowly and with the same cables and a charger putting on 5.3V - that .3v would increase the maximum charge rate by something on the order of 50%. So by force of consumer demand, almost everything that says it puts out 5v onto USB port, it usually actually supplies 5.2-5.5. Otherwise people give it crap reviews and say it's slow to charge. 

And yeah you always need to use a voltage divider. You cannot apply a voltage higher that the limit shown in the table posted above. The voltage dividers it has internally are only for VDD

**Comment by Brendy1986** [comment]
Thank you for answer

**Comment by biomurph** [comment]
Hi,
I figured I would ask this question here, rather than posting a new thread.
I am using a 1606, and I want to measure the voltage of the battery.
All my logic is running at 3.0V (ATtiny, accelerometer, USB-UART).
If I use analogReference() to set it to the 4.3 internal reference, will I be able to directly measure the battery voltage on one of the analog pins? I don't have a circuit lashed up to test in hardware yet, but I want to know if I'm on the right track?
Thanks for all the great work on this core!

**Comment by hmeijdam** [comment]
4.3V internal reference from a 3.0 V power supply? That's not going to work.

See below what the minimum supply voltages are to make the VREF work.
![image](https://github.com/user-attachments/assets/64f9a179-bb60-49cc-8fff-612769f8579a)

**Comment by biomurph** [comment]
Ah, I see. Thank you.
I guess I will have to use a V divider on the battery supply to test it.
I'll look into using a mosfet to only run the V divider when I want to look at it.

**Reply by grandaspanna** [reply]
You can flip your logic here. Use your Vdd supply as the reference and configure the ADC to read VREF (setup VREF as 2.048 or similar) as its input. Some simple maths will calculate Vdd.  This approach is described elsewhere and I’m pretty sure this will work on the 1606 (someone correct me if I’m wrong).

**Reply by jvasileff** [reply]
The code for that is actually in the first post of this thread.

It also provided as an example for the core - https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/libraries/megaTinyCore/examples/readTempVcc/readTempVcc.ino
